### PR TITLE
docs: add more notes from The Dream Machine

### PR DESCRIPTION
**Pull request by katcipis**
As I wrote this...it did give me a feeling that overall bottom-up would be better...so in the end Ken Thompson + @i4ki are indeed geniuses and my top-down brain is limited and stupid :-(

### CODE REVIEW

**Review by vitorarins — COMMENTED** [review]
Nice! I don't know why I thought it was the other way around. Like if you understand the problem  well enough you know exactly what will be the "bottom" parts. But if you just have an outline of it, would be easier to start with the "top", which in theory should be smaller (thinking about a pyramid 😛 )

File: notes/books/dream-machine.md [hunk 1]
```diff
@@ -40,3 +40,33 @@ depicted a math professor that made his name by stealing ideas
...
+For problems that are less understood, or that you have less understanding
+about the solution structure, bottom-up will be beneficial, since with
+bottom-up you focus on defining useful/small core blocks that you are
+not sure yet how you are going to compose to solve the final problem.
+The small blocks + composition allows you to explore more freely and
+iterate faster.
+
+On top of this overall idea, which I found intriguing/interesting, it was
+also defended that Lisp was well suited for bottom-up design, it is fairly
+easily to start with small functions and compose them into larger ones
```
Comment (i4ki) [inline notes/books/dream-machine.md:70]:
```suggestion
easy to start with small functions and compose them into larger ones
```

**Review by i4ki — APPROVED** [review]
lgtm

**Comment by i4ki** [comment]
Not at all... the exact opposite of genius because I can't handle complex stuff in my mind, so I solve the bottom part first because it's a constrained problem, the image/domain is smaller, and so on, so when it's done I create layers on top of this until I solve the actual problem. The issue is that I could have solved the bottom part completely wrong and now I have to go back and fix it... That's why this method also needs a lot of experimentation.

**Comment by katcipis** [comment]
> Nice! I don't know why I thought it was the other way around. Like if you understand the problem well enough you know exactly what will be the "bottom" parts. But if you just have an outline of it, would be easier to start with the "top", which in theory should be smaller (thinking about a pyramid stuck_out_tongue )

We ended up talking about this in person, I had a similar feeling but then I realized that starting at the "top" means you actually know what the top is, the idea that you need bottom/up smaller components first is when you don't know what top is, it is extreme exploration mode.

The scenario you described is:

1 - Problem is clear, you can think of an API 
2 - You have no idea how to design internally, which modules you are going to have, how to break up the problem

On that case I also usually go top/down, start with testing on the final API, and then I can refactor freely on internal modules, splitting, merging...etc. But I think he means this scenario:

1 - You have no idea exactly what the problem is, or a very vague idea
2 - You need to start building something, to get ideas, and explore, but again..no idea what the final API could be
3 - Start with very small components ?

Kinda like you just know that your problem is around image processing and you build a few building blocks, bottom up, around image processing, but you are not sure where are you going to arrive in the end =P.

**Comment by katcipis** [comment]
> Not at all... the exact opposite of genius because I can't handle complex stuff in my mind, so I solve the bottom part first because it's a constrained problem, the image/domain is smaller, and so on, so when it's done I create layers on top of this until I solve the actual problem. The issue is that I could have solved the bottom part completely wrong and now I have to go back and fix it... That's why this method also needs a lot of experimentation.

Maybe thinking you are not a genius and actually quite stupid is what makes you genius :exploding_head:

**Pull request merged**